src/ir/module-utils.h

```diff
@@ -466,8 +467,36 @@ collectSignatures(Module& wasm,
     }
   }
 
-  // TODO: recursively traverse each reference type, which may have a child type
-  //       this is itself a reference type.
+  // Recursively traverse each reference type, which may have a child type that
+  // is itself a reference type. This reflects an appearance in the binary
+  // format that is in the type section itself.
+  // As we do this we may find more and more signatures, as nested children of
+  // previous ones. Each such signature will appear in the type section once, so
+  // we just need to visit it once.
+  // TODO: handle struct and array fields
+  std::unordered_set<Signature> newSigs;
+  for (auto& pair : counts) {
+    newSigs.insert(pair.first);
+  }
+  while (!newSigs.empty()) {
+    auto iter = newSigs.begin();
+    auto sig = *iter;
+    newSigs.erase(iter);
```

Comment:
It would be slightly more efficient to pop from the back here.

Reply:
Is it possible? It is an `unordered_set`..

Reply:
Oops, ignore me 😬